include/libcyphal/platform/single_threaded_executor.hpp

```diff
@@ -0,0 +1,417 @@
+
+#include <algorithm>
+#include <chrono>
+#include <cstdint>
+#include <tuple>
+#include <utility>
+
+namespace libcyphal
+{
+namespace platform
+{
+
+/// @brief Defines platform agnostic single-threaded executor.
+///
+/// NOSONAR cpp:S4963 for below `class SingleThreadedExecutor` - we do directly handle resources here;
+/// namely: in destructor we de-allocation of any potentially left callback nodes we allocated previously.
+///
+class SingleThreadedExecutor : public IExecutor  // NOSONAR cpp:S4963
+{
+public:
+    explicit SingleThreadedExecutor(cetl::pmr::memory_resource& memory_resource)
+        : nodes_allocator_{&memory_resource}
+    {
+    }
+
+    ~SingleThreadedExecutor() override
+    {
+        // Just in case release whatever callback nodes left, but properly used `Callback::Handle`-s
+        // (aka "handle must not outlive executor") should have removed them all.
+        //
+        CETL_DEBUG_ASSERT(scheduled_nodes_.empty(), "");
+        CETL_DEBUG_ASSERT(registered_nodes_.empty(), "");
+        releaseAllCallbackNodes();
+    }
+
+    SingleThreadedExecutor(const SingleThreadedExecutor&)                = delete;
+    SingleThreadedExecutor(SingleThreadedExecutor&&) noexcept            = delete;
+    SingleThreadedExecutor& operator=(const SingleThreadedExecutor&)     = delete;
+    SingleThreadedExecutor& operator=(SingleThreadedExecutor&&) noexcept = delete;
+
+    struct SpinResult
+    {
+        /// Time of the next scheduled callback to execute,
+        /// or `cetl::nullopt` if there are no scheduled callbacks.
+        /// This can be used to let the application sleep/poll when there are no callbacks pending.
+        cetl::optional<TimePoint> next_exec_time;
+
+        /// An approximation of the maximum lateness observed during the spin call
+        /// (the real slack may be worse than the approximation).
+        /// This is always non-negative.
+        Duration worst_lateness;
+    };
+
+    CETL_NODISCARD SpinResult spinOnce()
+    {
+        SpinResult spin_result{};
+
+        auto approx_now = TimePoint::min();
+
+        while (auto* const scheduled_node = scheduled_nodes_.min())
+        {
+            // No linting b/c we know for sure the type of the node.
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-static-cast-downcast)
+            auto& callback_node = static_cast<CallbackNode&>(*scheduled_node);
+
+            const auto exec_time = callback_node.executionTime();
+            if (approx_now < exec_time)
+            {
+                approx_now = now();
+                if (approx_now < exec_time)
+                {
+                    spin_result.next_exec_time = exec_time;
+                    break;
+                }
+            }
+
+            spin_result.worst_lateness = std::max(approx_now - exec_time, spin_result.worst_lateness);
+
+            const bool is_removed = applyScheduleOnNextCallback(callback_node, exec_time);
+
+            callback_node(approx_now);
```

Comment:
What happens if the callback, or something in this callback's callstack, calls removeCallback on the executor? Do we handle this properly and do we cover this in unit test? What about if this callback calls removeCallback on the next scheduled node?

Reply:
Yes, we expect and handle this - callback will be removed only once, and any future attempts to remove won't find such (already removed) callback id in the tree, so no operation will be done.

Yes, there are 3 unit tests for such cases:
- `reset_once_scheduling_from_callback`
- `reset_once_scheduling_from_auto_remove_callback`
- `reset_repeat_scheduling_from_callback`